opentelemetry-exporter-geneva/geneva-uploader/src/payload_encoder/otlp_encoder.rs

```diff
@@ -0,0 +1,410 @@
+const FIELD_SEVERITY_TEXT: &str = "SeverityText";
+const FIELD_BODY: &str = "body";
+
+/// Encoder to write OTLP payload in bond form.
+#[derive(Clone)]
+pub struct OtlpEncoder {
+    // TODO - limit cache size or use LRU eviction, and/or add feature flag for caching
+    schema_cache: SchemaCache,
+}
+
+impl OtlpEncoder {
+    pub fn new() -> Self {
+        OtlpEncoder {
+            schema_cache: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    pub fn encode_log_batch<'a, I>(
+        &self,
+        logs: I,
+        metadata: &str,
+    ) -> Vec<(u64, String, Vec<u8>, usize)>
+    where
+        I: Iterator<Item = &'a opentelemetry_proto::tonic::logs::v1::LogRecord>,
+    {
+        use std::collections::HashMap;
+
+        let mut batches: LogBatches = HashMap::new();
+
+        for log_record in logs {
+            // 1. Get schema
+            let field_specs = self.determine_fields(log_record);
+            let schema_id = Self::calculate_schema_id(&field_specs);
+            let (schema_entry, field_info) = self.get_or_create_schema(schema_id, field_specs);
+
+            // 2. Encode row
+            let row_buffer = self.write_row_data(log_record, &field_info);
+            let event_name = if log_record.event_name.is_empty() {
+                "Log".to_string()
+            } else {
+                log_record.event_name.clone()
+            };
+            let level = log_record.severity_number as u8;
+
+            // 3. Insert into batches - Key is (schema_id, event_name)
+            batches
+                .entry((schema_id, event_name.clone()))
+                .or_insert_with(|| (schema_entry, Vec::new()))
+                .1
+                .push((event_name, level, row_buffer));
+        }
+
+        // 4. Encode blobs (one per schema AND event_name combination)
+        let mut blobs = Vec::new();
+        for ((schema_id, batch_event_name), (schema_entry, records)) in batches {
+            let events: Vec<CentralEventEntry> = records
+                .into_iter()
+                .map(|(event_name, level, row_buffer)| CentralEventEntry {
+                    schema_id,
+                    level,
+                    event_name,
+                    row: row_buffer,
+                })
+                .collect();
+            let events_len = events.len();
+
+            let blob = CentralBlob {
+                version: 1,
+                format: 2,
+                metadata: metadata.to_string(),
+                schemas: vec![schema_entry],
+                events,
+            };
+            let bytes = blob.to_bytes();
+            blobs.push((schema_id, batch_event_name, bytes, events_len));
+        }
+        blobs
+    }
+
+    /// Determine which fields are present in the LogRecord
+    fn determine_fields(&self, log: &LogRecord) -> Vec<FieldDef> {
+        // Pre-allocate with estimated capacity to avoid reallocations
+        let estimated_capacity = 4 + 6 + log.attributes.len();
```

Comment:
There are 7 entries for Part A and 4 for Part B

```suggestion
        let estimated_capacity = 7 + 4 + log.attributes.len();
```